pkg/storage/utils/decomposedfs/tree/propagator/sync.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2018-2021 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package propagator
+
+import (
+	"context"
+	"errors"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/cs3org/reva/v2/pkg/appctx"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/lookup"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/metadata"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/metadata/prefixes"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/node"
+	"github.com/rogpeppe/go-internal/lockedfile"
+)
+
+type SyncPropagator struct {
+	treeSizeAccounting bool
+	treeTimeAccounting bool
+	lookup             lookup.PathLookup
+}
+
+func NewSyncPropagator(treeSizeAccounting, treeTimeAccounting bool, lookup lookup.PathLookup) SyncPropagator {
+	return SyncPropagator{
+		treeSizeAccounting: treeSizeAccounting,
+		treeTimeAccounting: treeTimeAccounting,
+		lookup:             lookup,
+	}
+}
+
+func (p SyncPropagator) Propagate(ctx context.Context, n *node.Node, sizeDiff int64) error {
```

Comment:
Can we reuse the `Propagate` method from the sync propagator in the async propagator? Or at least extract some parts of it to a common function. I can see lots of duplication in these two methods....

Reply:
Phew... I just tried but there are lots of subtle changes and slightly different dependencies. I would suggest we defer this refactoring to a later point in time so we can get the feature in for testing now.